File: frontend/src/types/tenant.ts
/**
 * Tenant types — aligned with backend TenantResponse (iam/dto/tenant_dto.go).
 *
 * Backend returns: id, name, slug, domain, settings (JSONB), status,
 *   subscription_tier, created_at, updated_at.
 */

export interface Tenant {
  id: string;
  name: string;
  slug: string;
  domain: string | null;
  status: TenantStatus;
  subscription_tier: SubscriptionTier;
  /** Opaque JSONB — may be empty `{}` for tenants created without settings. */
  settings: TenantSettings;
  created_at: string;
  updated_at: string;
}

/**
 * Backend enum: active, inactive, suspended, trial, onboarding, deprovisioned.
 * See model/tenant.go TenantStatus constants.
 */
export type TenantStatus =
  | 'active'
  | 'inactive'
  | 'suspended'
  | 'trial'
  | 'onboarding'
  | 'deprovisioned';

/**
 * Backend enum: free, starter, professional, enterprise.
 * See model/tenant.go SubscriptionTier constants.
 */
export type SubscriptionTier = 'free' | 'starter' | 'professional' | 'enterprise';

/**
 * Settings is stored as JSONB in the tenants table.
 * All fields are optional because older tenants may have been created
 * with an empty `{}` settings object.
 */
export interface TenantSettings {
  max_users?: number;
  max_storage_gb?: number;
  enabled_suites?: string[];
  mfa_required?: boolean;
  session_timeout_minutes?: number;
  password_policy?: TenantPasswordPolicy;
  ip_whitelist?: string[];
  custom_domain?: string | null;
  branding?: TenantBranding;
  /** Organization-wide starting dashboard; users can still personalize it. */
  dashboard_defaults?: import('@/components/dashboard/widget-board/layout-utils').PersistedUserBoard;
}

export interface TenantBranding {
  logo_url: string | null;
  primary_color: string;
  accent_color: string;
  company_name: string;
}

export interface TenantPasswordPolicy {
  min_length: number;
  require_uppercase: boolean;
  require_lowercase: boolean;
  require_numbers: boolean;
  require_special: boolean;
  max_age_days: number;
  history_count: number;
}

/**
 * Provision request — maps to backend ProvisionTenantRequest.
 * Creates tenant + owner user with tenant-admin role.
 */
export interface ProvisionTenantRequest {
  name: string;
  slug: string;
  subscription_tier: SubscriptionTier;
  owner_email: string;
  owner_name: string;
  /** Optional. If omitted the backend generates a random password returned in the response. */
  owner_password?: string;
  settings?: Partial<TenantSettings>;
}

/** Returned by POST /api/v1/tenants/provision — includes the one-time temp_password. */
export interface ProvisionTenantResponse extends Tenant {
  temp_password: string;
}

export interface TenantUsage {
  tenant_id: string;
  period: string;
  active_users: number;
  api_calls: number;
  storage_used_bytes: number;
  bandwidth_bytes: number;
  suite_usage: Record<string, SuiteUsage>;
}

export interface SuiteUsage {
  suite: string;
  api_calls: number;
  active_users: number;
  last_accessed: string | null;
}

File: backend/internal/iam/model/tenant.go
package model

import (
	"encoding/json"
	"fmt"
	"net"
	"time"
)

type TenantStatus string

const (
	TenantStatusActive        TenantStatus = "active"
	TenantStatusInactive      TenantStatus = "inactive"
	TenantStatusSuspended     TenantStatus = "suspended"
	TenantStatusTrial         TenantStatus = "trial"
	TenantStatusOnboarding    TenantStatus = "onboarding"
	TenantStatusDeprovisioned TenantStatus = "deprovisioned"
)

type SubscriptionTier string

const (
	TierFree         SubscriptionTier = "free"
	TierStarter      SubscriptionTier = "starter"
	TierProfessional SubscriptionTier = "professional"
	TierEnterprise   SubscriptionTier = "enterprise"
)

type Tenant struct {
	ID               string
	Name             string
	Slug             string
	Domain           *string
	Settings         json.RawMessage
	Status           TenantStatus
	SubscriptionTier SubscriptionTier
	DeprovisionedAt  *time.Time
	DeprovisionedBy  *string
	RetainUntil      *time.Time
	CreatedAt        time.Time
	UpdatedAt        time.Time
}

// TenantSettings is the typed representation of the Settings JSONB column.
// All fields are optional (existing tenants may have empty settings).
type TenantSettings struct {
	MaxUsers              *int              `json:"max_users,omitempty"`
	MaxStorageGB          *int              `json:"max_storage_gb,omitempty"`
	EnabledSuites         []string          `json:"enabled_suites,omitempty"`
	MFARequired           *bool             `json:"mfa_required,omitempty"`
	SessionTimeoutMinutes *int              `json:"session_timeout_minutes,omitempty"`
	PasswordPolicy        *PasswordPolicy   `json:"password_policy,omitempty"`
	IPWhitelist           []string          `json:"ip_whitelist,omitempty"`
	CustomDomain          *string           `json:"custom_domain,omitempty"`
	Branding              *BrandingSettings `json:"branding,omitempty"`
}

// PasswordPolicy defines tenant-level password rules.
type PasswordPolicy struct {
	MinLength        int  `json:"min_length"`
	RequireUppercase bool `json:"require_uppercase"`
	RequireLowercase bool `json:"require_lowercase"`
	RequireNumbers   bool `json:"require_numbers"`
	RequireSpecial   bool `json:"require_special"`
	MaxAgeDays       int  `json:"max_age_days"`
	HistoryCount     int  `json:"history_count"`
}

// BrandingSettings defines tenant UI customization.
type BrandingSettings struct {
	LogoURL      *string `json:"logo_url"`
	PrimaryColor string  `json:"primary_color"`
	AccentColor  string  `json:"accent_color"`
	CompanyName  string  `json:"company_name"`
}

// validSuites is the allowed set of suite identifiers.
var validSuites = map[string]bool{
	"cyber": true, "data": true, "acta": true,
	"lex": true, "visus": true, "platform": true,
}

// ValidateSettingsJSON validates the structure and field values of a tenant settings JSON blob.
// Returns nil if the settings are valid or empty.
func ValidateSettingsJSON(raw json.RawMessage) error {
	if len(raw) == 0 {
		return nil
	}

	var s TenantSettings
	if err := json.Unmarshal(raw, &s); err != nil {
		return fmt.Errorf("invalid settings JSON: %w", err)
	}

	if s.MaxUsers != nil && *s.MaxUsers < 0 {
		return fmt.Errorf("max_users must be non-negative")
	}
	if s.MaxStorageGB != nil && *s.MaxStorageGB < 0 {
		return fmt.Errorf("max_storage_gb must be non-negative")
	}
	if s.SessionTimeoutMinutes != nil && *s.SessionTimeoutMinutes < 1 {
		return fmt.Errorf("session_timeout_minutes must be at least 1")
	}
	for _, suite := range s.EnabledSuites {
		if !validSuites[suite] {
			return fmt.Errorf("invalid suite %q in enabled_suites", suite)
		}
	}
	for _, ip := range s.IPWhitelist {
		if net.ParseIP(ip) == nil {
			// Try CIDR
			if _, _, err := net.ParseCIDR(ip); err != nil {
				return fmt.Errorf("invalid IP address or CIDR %q in ip_whitelist", ip)
			}
		}
	}
	if s.PasswordPolicy != nil {
		if s.PasswordPolicy.MinLength < 0 {
			return fmt.Errorf("password_policy.min_length must be non-negative")
		}
		if s.PasswordPolicy.MaxAgeDays < 0 {
			return fmt.Errorf("password_policy.max_age_days must be non-negative")
		}
		if s.PasswordPolicy.HistoryCount < 0 {
			return fmt.Errorf("password_policy.history_count must be non-negative")
		}
	}

	return nil
}
